samples/foc_controller/src/foc_driver.c

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) 2022 Linaro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <arm_math.h>
+#include <string.h>
+#include <zephyr.h>
+#include <device.h>
+#include <drivers/i2c.h>
+#include <drivers/pwm.h>
+#include <drivers/gpio.h>
+#include <sys/printk.h>
+#include <step/sample_pool.h>
+#include <step/proc_mgr.h>
+#include "foc_driver.h"
+
+/** AS5600 I2C Encoder constants */
+#define AS5600_SLAVE_ADDR 0x36
+#define AS5600_ANGLE_REGISTER_H 0x0E
+#define AS5600_PULSES_PER_REVOLUTION 4096.0f
+#define AS5600_READING_MASK 0xFFF 
+
+#define ROTOR_DRV_PAYLOAD_SZ      (sizeof(struct foc_controller_payload))
+#define ROTOR_DRV_PAYLOAD_SRC_ID  (2)
+
+/* PWM frequency about 16KHz */
+#define PWM_PERIOD_NSEC 62500
+
+/**
+ * @brief Pre-populated measurement header to copy into allocated measurements.
+ */
+static struct step_mes_header rotor_sensor_header = {
+	/* Filter word. */
+	.filter = {
+		.base_type = STEP_MES_TYPE_PHASE,
+		.flags = {
+			.data_format = STEP_MES_FORMAT_NONE,
+			.encoding = STEP_MES_ENCODING_NONE,
+			.compression = STEP_MES_COMPRESSION_NONE,
+			.timestamp = STEP_MES_TIMESTAMP_UPTIME_MS_32,
+		},
+	},
+	/* SI Unit word. */
+	.unit = {
+		.si_unit = STEP_MES_UNIT_SI_RADIAN,
+		.ctype = STEP_MES_UNIT_CTYPE_IEEE754_FLOAT32,
+		.scale_factor = STEP_MES_SI_SCALE_NONE,
+	},
+	/* Source/Len word. */
+	.srclen = {
+		.len = ROTOR_DRV_PAYLOAD_SZ,
+		.fragment = 0,
+		.sourceid = ROTOR_DRV_PAYLOAD_SRC_ID,
+	}
+};
+
+static void foc_driver_rotor_position_sample_thread(void *arg);
+
+/* Rotor sensor measurement thread. */
+K_THREAD_DEFINE(step_pm_tid, 4096,
+		foc_driver_rotor_position_sample_thread, NULL, NULL, NULL,
+		-1, 0, 0);
+
+const static float encoder_to_degrees_ratio = (360.0f) / AS5600_PULSES_PER_REVOLUTION;
+
+static foc_measurement_callback_t user_callback;
+static float bus_dc_link_voltage;
+static float bus_dc_link_bias_voltage;
+static float voltage_to_duty_ratio;
+static float pole_pairs;
+
+static struct {
+	uint16_t raw;
+    uint16_t zero_offset;
```

Comment:
Minor nit, but inconsistent leading whitespace:

```suggestion
	uint16_t raw;
	uint16_t zero_offset;
```